server/src/main/java/org/apache/druid/catalog/model/table/FormattedInputSourceDefn.java

```diff
@@ -95,7 +95,10 @@ protected List<ParameterDefn> addFormatParameters(
   )
   {
     final List<ParameterDefn> toAdd = new ArrayList<>();
-    final ParameterDefn formatProp = new Parameter(FORMAT_PARAMETER, ParameterType.VARCHAR, false);
+    // While the format parameter is required, we mark it as optional. Else
+    // if the source defines optional parameters, they will still be ignored
```

Comment:
if the format* defines optional parameters?

Reply:
Actually, _input source_, as written. The input source parameters preceded the input format parameters. That's how we end up with the scenario in the comment: the input source has two optional parameters, but the format has a required one. That forces the input source parameters to also become required.

This is, by the way, why that bug in `extern`, where I had the optional flag backward, actually worked. Calcite took my `(optional, optional, required)` and turned it into what the code originally had: `(required, required, required)`.